Agda fails to compile code involving @0

Agda fails to compile the following code properly:
```agda
module Bug where

record Erased (A : Set) : Set where
  constructor [_]
  field
    @0 erased : A

open Erased

record R (M : Set → Set) : Set₁ where
  field
    f : (A B : Set) → M A → (A → M B) → M B

works : R Erased
R.f works _ _ x g .erased = erased (g (erased x))

fails : R Erased
R.f fails _ _ x g = [ erased (g (erased x)) ]
```
GHC complains:
```
[…]/MAlonzo/Code/Bug.hs:31:50: error:
    • Variable not in scope: d10
    • Perhaps you meant one of these: ‘d16’ (line 16), ‘d30’ (line 22)
   |
31 | d38 = coe C47 (coe (\ v0 v1 v2 v3 -> coe v3 (coe d10 erased)))
   |                                                  ^^^
```
The generated Haskell code:
```haskell
{-# LANGUAGE EmptyDataDecls, EmptyCase, ExistentialQuantification,
             ScopedTypeVariables, NoMonomorphismRestriction, RankNTypes,
             PatternSynonyms #-}
module MAlonzo.Code.Bug where

import MAlonzo.RTE (coe, erased, AgdaAny, addInt, subInt, mulInt,
                    quotInt, remInt, geqInt, ltInt, eqInt, eqFloat, add64, sub64,
                    mul64, quot64, rem64, lt64, eq64, word64FromNat, word64ToNat)
import qualified MAlonzo.RTE
import qualified Data.Text

name4 = "Bug.Erased"
d4 a0 = ()
data T4 = C12
name16 = "Bug.R"
d16 a0 = ()
newtype T16
  = C47 (() -> () -> AgdaAny -> (AgdaAny -> AgdaAny) -> AgdaAny)
name30 = "Bug.R.f"
d30 ::
  T16 -> () -> () -> AgdaAny -> (AgdaAny -> AgdaAny) -> AgdaAny
d30 v0
  = case coe v0 of
      C47 v1 -> coe v1
      _ -> MAlonzo.RTE.mazUnreachableError
name32 = "Bug.works"
d32 :: T16
d32 = coe C47 erased
name38 = "Bug.fails"
d38 :: T16
d38 = coe C47 (coe (\ v0 v1 v2 v3 -> coe v3 (coe d10 erased)))
```
I'm guessing that `d10` corresponds to the projection `erased`, which is not compiled:
```
Not compiling Erased.erased.
```

The following code exhibits a similar problem:
```agda
open import Agda.Builtin.Equality
open import Agda.Builtin.Sigma

P : (A B : Set) → (A → B) → Set
P A B f = (y : B) → Σ A (λ x → f x ≡ y)

record R (A B : Set) : Set where
  field
    to      : A → B
    from    : B → A
    from-to : (x : A) → from (to x) ≡ x

p : (A B : Set) (r : R A B) → P A B (R.to r)
p _ _ r y = R.from r y , to-from y
  where
  postulate
    to-from : ∀ x → R.to r (R.from r x) ≡ x

record Erased (A : Set) : Set where
  constructor box
  field
    @0 erased : A

p′ : (A : Set) → P (Erased A) (Erased (Erased A)) (box {A = Erased A})
p′ A = p _ _ (record
  { to      = box {A = _}
  ; from    = λ { (box (box x)) → box {A = A} x }
  ; from-to = λ x → refl {A = Erased A} {x = _}
  })
```
Error message:
```
[…]/MAlonzo/Code/Bug.hs:52:34: error:
    • Variable not in scope: d72
    • Perhaps you meant one of these:
        ‘MAlonzo.Code.Agda.Builtin.Equality.d12’ (imported from MAlonzo.Code.Agda.Builtin.Equality),
        ‘d22’ (line 18), ‘d42’ (line 34)
   |
52 | du78 = coe du50 (coe C179 erased d72)
   |                                  ^^^
```
Here `d72` corresponds to `erased`.

Agda generates the following code:
```haskell
du78 :: T66 -> MAlonzo.Code.Agda.Builtin.Sigma.T14
du78 = coe du50 (coe C179 erased d72)
```
The code is compiled without complaints if the first underscore in `p′` is replaced by `Erased A`, or if the second underscore is replaced by `Erased (Erased A)`, or if the third underscore is replaced by `Erased A`, or if the fourth underscore is replaced by `x`. In either of these cases Agda generates the following code:
```haskell
du78 :: T66 -> MAlonzo.Code.Agda.Builtin.Sigma.T14
du78 = coe du50 (coe C179 erased erased)
```

--- a/src/full/Agda/TypeChecking/MetaVars.hs
+++ b/src/full/Agda/TypeChecking/MetaVars.hs
@@ -1299,12 +1299,12 @@ inverseSubst args = map (mapFst unArg) <$> loop (zip args terms)
           let fallback
                | isIrrelevant info = return vars
                | otherwise                              = failure
-          isRC <- lift $ isRecordConstructor $ conName c
           irrProj <- optIrrelevantProjections <$> pragmaOptions
-          case isRC of
+          (lift $ isRecordConstructor $ conName c) >>= \case
             Just (_, r@Record{ recFields = fs })
-              | YesEta <- recEtaEquality r  -- Andreas, 2019-09-10, issue #4185: only for eta-records
+              | YesEta <- recEtaEquality r  -- Andreas, 2019-11-10, issue #4185: only for eta-records
               , length fs == length es
+              , all usableQuantity fs       -- Andreas, 2019-11-12b, issue #4168b
               , irrProj || all isRelevant fs -> do
                 let aux (Arg _ v) (Arg info' f) = (Arg ai v,) $ t `applyE` [Proj ProjSystem f] where
                      ai = ArgInfo

--- a/test/Compiler/simple/Issue4168-4185.agda
+++ b/test/Compiler/simple/Issue4168-4185.agda
@@ -1,0 +1,69 @@
+-- Andreas, 2019-11-12, issue #4168b
+--
+-- Meta variable solver should not use uncurrying
+-- if the record type contains erased fields.
+
+open import Agda.Builtin.Equality
+open import Agda.Builtin.Sigma
+open import Common.IO
+
+P : (A B : Set) → (A → B) → Set
+P A B f = (y : B) → Σ A (λ x → f x ≡ y)
+
+record R (A B : Set) : Set where
+  field
+    to      : A → B
+    from    : B → A
+    from-to : (x : A) → from (to x) ≡ x
+
+p : (A B : Set) (r : R A B) → P A B (R.to r)
+p _ _ r y = R.from r y , to-from y
+  where
+  postulate
+    to-from : ∀ x → R.to r (R.from r x) ≡ x
+
+record Box (A : Set) : Set where
+  constructor box
+  field
+    @0 unbox : A
+
+test : (A : Set) → P (Box A) (Box (Box A)) (box {A = Box A})
+test A = p _ _ (record
+  { to      = box {A = _}
+  ; from    = λ { (box (box x)) → box {A = A} x }
+      -- at first, a postponed type checking problem ?from
+  ; from-to = λ x → refl {A = Box A} {x = _}
+  })
+-- from-to creates constraint
+--
+--   x : Box A |- ?from (box x) = x : Box A
+--
+-- This was changed to
+--
+--   y : Box (Box A) |- ?from y = unbox y : Box A
+--
+-- which is an invalid transformation since x is not
+-- guaranteed to be in erased context.
+
+-- As a consequence, compilation (GHC backend) failed.
+
+
+-- A variant with a non-erased field.
+
+record ⊤ : Set where
+
+record Box' (A : Set) : Set where
+  constructor box'
+  field
+    unit : ⊤
+    @0 unbox' : A
+
+test' : (A : Set) → P (Box' A) (Box' (Box' A)) (box' {A = Box' A} _)
+test' A = p _ _ (record
+  { to      = box' {A = _} _
+  ; from    = λ { (box' _ (box' _ x)) → box' {A = A} _ x }
+  ; from-to = λ x → refl {A = Box' A} {x = _}
+  })
+
+main : IO ⊤
+main = return _

--- a/test/Compiler/simple/Issue4168-4185.out
+++ b/test/Compiler/simple/Issue4168-4185.out
@@ -1,0 +1,3 @@
+EXECUTED_PROGRAM
+
+ret > ExitSuccess

